### PR TITLE
Buffer python warning to avoid deadlocks

### PR DESCRIPTION
**Pull request by albanD**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #27154 Handle warning in torchscript
* #26614 add pybind version of HANDLE_TH_ERRORS
* **#26613 Buffer python warning to avoid deadlocks**
* #28283 Fix clang-tidy errors in csrc/Module.cpp

Differential Revision: [D18249633](https://our.internmc.facebook.com/intern/diff/D18249633)

### CODE REVIEW

**Comment by ezyang** [comment]
You need to rebase to master to fix build failures

File: torch/csrc/Exceptions.h [hunk 1]
```diff
@@ -138,4 +157,27 @@ struct ValueError : public PyTorchError {
   }
 };
 
+// ATen warning handler for Python
```
Comment (ezyang) [inline torch/csrc/Exceptions.h]: I see this code was moved. Any substantive changes inside?

File: torch/csrc/Exceptions.h [hunk 2]
```diff
@@ -138,4 +157,27 @@ struct ValueError : public PyTorchError {
...
+struct PyWarningHandler{
+public:
+  using warning_buffer_t =
+    std::queue<std::pair<c10::SourceLocation, std::string>>;
+
+  static void py_warning_handler(
+    const c10::SourceLocation& source_location,
+    const std::string& msg);
+
+  static warning_buffer_t warning_buffer;
```
Comment (ezyang) [inline torch/csrc/Exceptions.h]: This struct is all static members. Any reason for it to be a struct? A namespace would have worked equally as well. You can move the `warning_buffer` declaration to be private to `torch/csrc/Exceptions.cpp` as it doesn't seem to be used anywhere else.

File: torch/csrc/Exceptions.h [hunk 3]
```diff
@@ -138,4 +157,27 @@ struct ValueError : public PyTorchError {
...
+};
+
+struct EnforceWarningBuffer {
+public:
+/// See NOTE [ Conversion Cpp Python Warning ] for noexcept justification
+  EnforceWarningBuffer() noexcept(true);
+  ~EnforceWarningBuffer() noexcept(false);
```
Comment (ezyang) [inline torch/csrc/Exceptions.h]: Wow, TIL this is allowed hahaha

File: torch/csrc/Exceptions.h [hunk 4]
```diff
@@ -2,38 +2,57 @@
...
 
 #include <torch/csrc/THP_export.h>
 #include <torch/csrc/utils/auto_gil.h>
 #include <torch/csrc/WindowsTorchApiMacro.h>
+#include "c10/util/StringUtil.h"
+#include "c10/util/Exception.h"
+
+/// NOTE [ Conversion Cpp Python Warning ]
+/// Python warning semantic is different from the cpp one in that
+/// they can raise errors. To handle this case which requires
```
Comment (ezyang) [inline torch/csrc/Exceptions.h]: I'm not sure this first sentence makes sense to anyone who didn't participate in the design discussion. An example would make it a lot clearer, I think.

Comment (ezyang) [reply torch/csrc/Exceptions.h]: Basically, walk the reader through the case when a warning handler raises an error, while the function itself is throwing an exception. (That diagram we drew, but in comments.)

File: torch/csrc/Exceptions.h (hunk 2 above)
Comment (ezyang) [reply torch/csrc/Exceptions.h]: Is `warning_buffer` thread safe? I don't think `std::queue` is thread safe. So what lock protects it?

File: torch/csrc/Exceptions.h (hunk 1 above)
Comment (albanD) [reply torch/csrc/Exceptions.h]: Yes only the comment was moved. The rest of the code is new !

File: torch/csrc/Exceptions.h (hunk 3 above)
Comment (albanD) [reply torch/csrc/Exceptions.h]: This is my attempt to make sure that we don't get killed by throwing two exceptions.

File: c10/util/Exception.cpp [hunk 5]
```diff
@@ -64,16 +64,20 @@ void Error::AppendMessage(const std::string& new_msg) {
 }
 
 void Warning::warn(SourceLocation source_location, std::string msg) {
-  warning_handler_(source_location, msg.c_str());
+  warning_handler_(source_location, msg);
```
Comment (ezyang) [inline c10/util/Exception.cpp]: nit: `std::move(msg)` here

Comment (ezyang) [reply c10/util/Exception.cpp]: Actually, it looks like the warning handler takes this by const reference. In that case, warn() should also take by const reference...

File: torch/csrc/Exceptions.cpp [hunk 6]
```diff
@@ -137,5 +137,73 @@ ValueError::ValueError(const char *format, ...) {
...
+    AutoGIL gil;
+
+    PyObject *ptype, *pvalue, *ptraceback;
+    PyErr_Fetch(&ptype, &pvalue, &ptraceback);
+
+    if(ptype) {
+      // A python error happened after the warning
+      // Simply handle with the cpp handler
+      while(warning_buffer.size() > 0) {
+        auto warning = warning_buffer.front();
```
Comment (ezyang) [inline torch/csrc/Exceptions.cpp]: `auto warning = std::move(warning_buffer.front());` here is better, it avoids an unnecessary copy assignment

Comment (ezyang) [reply torch/csrc/Exceptions.cpp]: But tbh, C++ `queue` is really bad and there isn't really a good reason to use it here: you can just stuff everything on a vector, then process the vector front to back and reset() it to empty when you're all done. (You don't need the ability to incrementally remove items; you only need to remove them all at once in the end.)

File: torch/csrc/Exceptions.cpp [hunk 7]
```diff
@@ -137,5 +137,73 @@ ValueError::ValueError(const char *format, ...) {
...
+        warning_buffer.pop();
+        auto source_location = warning.first;
+        auto msg = processErrorMsg(warning.second);
+        c10::Warning::warn(source_location, msg);
```
Comment (ezyang) [inline torch/csrc/Exceptions.cpp]: The use of `warn` here suggests to me that there is no good reason not to pass the `std::string` in by value all the way, as the caller never needs to retain `msg` in the end.

**Comment by ezyang** [comment]
Looks good. Just some C++ bits; I also wanna know what the locking strategy is.

**Comment by ezyang** [comment]
Don't forget to request re-review (click the little circle arrow icon next to the reviewer) when you want rereview

**Comment by albanD** [comment]
@ezyang  Clang-tidy errors are unrelated to my changes, can I ignore it?

**Comment by ezyang** [comment]
Yes, please don't fix clang tidy in this PR.

On second review, I'm also feeling like there is a missing comment saying *why* we are going through all of this rigamarole in the first place. Obviously we have had the design discussion in person, but none of that design discussion has been replicated, neither in the code, or the PR body.

File: torch/csrc/Exceptions.cpp [hunk 8]
```diff
@@ -137,5 +137,81 @@ ValueError::ValueError(const char *format, ...) {
...
+
+/// See NOTE [ Conversion Cpp Python Warning ] for noexcept justification
+/// NOLINTNEXTLINE(bugprone-exception-escape)
+EnforceWarningBuffer::~EnforceWarningBuffer() noexcept(false) {
+  c10::Warning::set_warning_handler(prev_handler);
+
+  std::unique_lock<std::mutex> lock(warning_buffer_mutex);
+
+  if(warning_buffer.size() > 0) {
+    AutoGIL gil;
```
Comment (ezyang) [inline torch/csrc/Exceptions.cpp]: Can `py_warning_handler` ever be called when we have the GIL? In that case, there's potential for deadlock here, right? One thread has the GIL and is waiting on warning buffer mutex, the other thread has warning buffer mutex and is waiting on the GIL.

**Review by ezyang — CHANGES_REQUESTED** [review]
deadlock

File: torch/csrc/Exceptions.cpp [hunk 9]
```diff
@@ -137,5 +137,89 @@ ValueError::ValueError(const char *format, ...) {
...
+
+/// See NOTE [ Conversion Cpp Python Warning ] for noexcept justification
+/// NOLINTNEXTLINE(bugprone-exception-escape)
+EnforceWarningBuffer::~EnforceWarningBuffer() noexcept(false) {
+  c10::Warning::set_warning_handler(prev_handler);
+
+  bool has_warnings;
+  {
+    std::unique_lock<std::mutex> lock(warning_buffer_mutex);
+    has_warnings = warning_buffer.size() > 0;
```
Comment (ezyang) [inline torch/csrc/Exceptions.cpp]: Hmmm. While it is certainly true that you avoid the deadlock this way, taking out a lock to test size, and then releasing the lock, and then taking the lock out later is definitely a concurrency antipattern. What if another thread modifies `warning_buffer` in-between the time you check and then when you actually run?

But this raises further questions in my mind: why is the warning buffer global (as opposed to something thread local?)

Comment (albanD) [reply torch/csrc/Exceptions.cpp]: Because it needs to be accessed from the static `at::warn()` function.
Is there a way to bind to bind to call `static void set_warning_handler(handler_t handler);` with a handler which is a function associated with this class instance? For example if I had a `EnforceWarningBuffer::warn_handler()` (non static function).

Comment (albanD) [reply torch/csrc/Exceptions.cpp]: It is not thread local mainly because that would force us to duplicate all the logic that currently exists for errors to move them from one thread to the other whenever multithreaded is used.

Comment (ezyang) [reply torch/csrc/Exceptions.cpp]: Well, if the warning handler takes a straight up function pointer there's nothing you can do, but if you change it to accept something more interesting you would be in better shape. Like it takes a pointer with a virtual method on it to handle.

The justification for not TLS feels flimsy to me, though I understand from a local perspective why it would be nicer to avoid having to solve the TLS transfer across threads problem. In particular, the reason why it feels dodgy to me is because it means one thread can queue warnings, and then they magically show up from another thread when returning to Python. Warnings will literally escape the scopes they were supposed to live in...

Comment (albanD) [reply torch/csrc/Exceptions.cpp]: After discussion. I'll update the PR to detect when such "escape" can happen. And warn the user in that case.

**Comment by albanD** [comment]
After more issues and more discussion. We're going to move this handler to a thread local variable.

File: c10/util/Exception.h [hunk 10]
```diff
@@ -82,26 +82,31 @@ class C10_API Error : public std::exception {
...
+/// warning handler.
+C10_API void warn(SourceLocation source_location, const std::string& msg);
+/// Sets the global warning handler. This is not thread-safe, so it should
+/// generally be called once during initialization or while holding the GIL
+/// for programs that use python.
+/// User is responsible for keeping the WarningHandler alive until
+/// it is not needed.
+C10_API void set_warning_handler(WarningHandler* handler) noexcept(true);
+/// Gets the global warning handler.
+C10_API WarningHandler* get_warning_handler() noexcept(true);
```
Comment (ezyang) [inline c10/util/Exception.h]: Nit: expose this functionality as an RAII guard so people don't forget to reset the warning handler when control exits, e.g., via an exception

Comment (ezyang) [reply c10/util/Exception.h]: I guess PyWarningHandler brings its own RAII thing.

**Review by ezyang — APPROVED** [review]

File: docs/cpp/source/Doxyfile [hunk 11]
```diff
@@ -70,6 +70,8 @@ RECURSIVE              = YES
 # Blacklist certain file patterns from the INPUT section.
 EXCLUDE = ../../../torch/csrc/api/include/torch/nn/pimpl-inl.h \
           ../../../torch/csrc/api/include/torch/detail
+# Increase the max node size for our large files
+DOT_GRAPH_MAX_NODES    = 100
```
Comment (ezyang) [inline docs/cpp/source/Doxyfile]: don't put this in an unrelated diff, this is what stacks are for!! Cross referencing with  https://github.com/pytorch/pytorch/pull/29041

Comment (albanD) [reply docs/cpp/source/Doxyfile]: This is actually related. the cpp docs does not build for Exception.h without this.
See build failures on the previous commit here: https://github.com/pytorch/pytorch/runs/284960182
